### PR TITLE
chore(run/jobs): add retries for flaky Cloud Builds

### LINKED ISSUE #8465: run.jobs.e2e_test: test_end_to_end failed
Note: #8211 was also for this test, but it was closed more than 10 days ago. So, I didn't mark it flaky.

----

commit: 2ca8165875b1d93fdee551bdaa4db410e50c8d85
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/0bf368db-0a83-4958-a9c5-2dd04e4a951e), [Sponge](http://sponge2/0bf368db-0a83-4958-a9c5-2dd04e4a951e)
status: failed
<details><summary>Test output</summary><br><pre>Traceback (most recent call last):
  File "/workspace/run/jobs/e2e_test.py", line 49, in setup_job
    "_SERVICE=" + SERVICE + ",_VERSION=" + SUFFIX + ",_REGION=" + REGION,
  File "/usr/local/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['gcloud', 'builds', 'submit', '--config', 'e2e_test_setup.yaml', '--project', 'python-docs-samples-tests', '--substitutions', '_SERVICE=job-quickstart-a6997be597,_VERSION=a6997be597,_REGION=us-west1']' returned non-zero exit status 1.</pre></details>

### PR DESCRIPTION
**Pull request by ace-n**
Fixes #8465

### CODE REVIEW

**Review by WildSunLove — APPROVED** [review]
LGTM

**Comment by davidcavazos** [comment]
Would [`backoff`](https://pypi.org/project/backoff) be able to handle this more cleanly? Since they're tests I don't have a strong opinion.

**Review by davidcavazos — APPROVED** [review]

File: run/jobs/e2e_test.py [hunk 1]
```diff
@@ -36,36 +36,61 @@
...
-            "e2e_test_setup.yaml",
-            "--project",
-            PROJECT,
-            "--substitutions",
-            "_SERVICE=" + SERVICE + ",_VERSION=" + SUFFIX + ",_REGION=" + REGION,
-        ]
-    )
+    # Retry up to 3 times
+    built = False
+    for x in range(3):
```
Comment (dandhlee) [inline run/jobs/e2e_test.py:41]: Like David suggested, would it be possible to use [`backoff`](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#retry-rpcs)?

File: run/jobs/e2e_test.py [hunk 2]
```diff
@@ -36,36 +36,61 @@
...
+                    "--config",
+                    "e2e_test_setup.yaml",
+                    "--project",
+                    PROJECT,
+                    "--substitutions",
+                    "_SERVICE=" + SERVICE + ",_VERSION=" + SUFFIX + ",_REGION=" + REGION,
+                ]
+            )
+            built = True
+        except Exception as e:
```
Comment (dandhlee) [inline run/jobs/e2e_test.py:57]: Are there specific exceptions that can be used, rather than generics?

File: run/jobs/e2e_test.py [hunk 3]
```diff
@@ -36,36 +36,61 @@
...
+            print(e)
+
```
Comment (dandhlee) [inline run/jobs/e2e_test.py:59]: Are there any cleanups that needs to be done, possibly in a `finally` clause?

File: run/jobs/e2e_test.py [hunk 4]
```diff
@@ -36,36 +36,61 @@
...
-            "--config",
-            "e2e_test_cleanup.yaml",
-            "--project",
-            PROJECT,
-            "--substitutions",
-            "_SERVICE=" + SERVICE + ",_VERSION=" + SUFFIX + ",_REGION=" + REGION,
-        ]
-    )
+    destroyed = False
+    for x in range(3):
```
Comment (dandhlee) [inline run/jobs/e2e_test.py:70]: Same 3 concerns as above! (backoff, exception, and cleanup using finally)

File: run/jobs/e2e_test.py (hunk 2 above)
Comment (ace-n) [reply run/jobs/e2e_test.py]: Done, via `backoff`

File: run/jobs/e2e_test.py (hunk 1 above)
Comment (ace-n) [reply run/jobs/e2e_test.py]: Done.

File: run/jobs/e2e_test.py (hunk 4 above)
Comment (ace-n) [reply run/jobs/e2e_test.py]: This step looks like the clean-up step for the previous one.

In any case - is a `finally` clause required if we're working with `backoff`?

File: run/jobs/e2e_test.py (hunk 3 above)
Comment (ace-n) [reply run/jobs/e2e_test.py]: See other thread. 🙂

File: run/jobs/e2e_test.py (hunk 4 above)
Comment (dandhlee) [reply run/jobs/e2e_test.py]: It's a bit tricky in this case actually - I'm not sure what would be the best approach if `teardown` were to throw errors. In other cases I'd suggest `finally` to clean up but fine to leave it as is here.

**Review by dandhlee — APPROVED** [review]
Thank you!

**Pull request merged**